### PR TITLE
Refactor cli/index.js

### PR DESCRIPTION
**Pull request by lhw-1**
**What is the purpose of this pull request?**

- [ ] Documentation update
- [ ] Bug fix
- [ ] Feature addition or enhancement
- [X] Code maintenance
- [ ] DevOps
- [x] Improve developer experience
- [ ] Others, please explain:

**Overview of changes:**

Refactors `cli/index.js` into multiple files, and separates it by command (`init`, `serve`, `build`, `deploy`).

**Anything you'd like to highlight/discuss:**

This refactor will encourage further modularity of our CLI and allow for easier development / debugging of the CLI commands and processes. The idea to do this was initially proposed in [my comment](https://github.com/MarkBind/markbind/issues/1756#issuecomment-1466383919) in #1756 to aid in the refactoring of `Site/index.js`, and was again mentioned in #2218.

A further refactor, specifically for the `serve` command, may be a possibility as well.

And of course, we should add more tests for each of the commands (in a separate PR).

**Testing instructions:**

Ensure that all the commands - `init`, `serve`, `build`, and `deploy` - are working as intended and as specified in the User Guide.

**Proposed commit message: (wrap lines at 72 characters)**

```
Refactor cli/index.js

The code for cli/index.js, especially for the serve command, is
becoming longer.

As more functionalities are added to the CLI commands, it is
becoming difficult to navigate the code.

Extracting the code for each command into a separate function
encourages modularity in developing the commands without 
worrying about coupling, allowing for a better debugging
experience and better code quality in general.

Let's extract all the code bodies for each command into a
separate file dedicated to a single command, and place all
common functions into the cli/src/util directory.

Doing this is preferable as we can prevent cli/index.js from
becoming too lengthy for a simple refactoring like Site/index.js.
```
---

**Checklist:** :ballot_box_with_check:

- [ ] Updated the documentation for feature additions and enhancements
- [ ] Added tests for bug fixes or features
- [X] Linked all related issues
- [X] No unrelated changes

### CODE REVIEW

**Review by yucheng11122017 — APPROVED** [review]
Thanks for the work @lhw-1! It definitely looks easier to read through now. 

I think serve could use more refactoring but will leave it to another PR. 

LGTM

File: packages/cli/index.js
Comment (lhw-1) [reply packages/cli/index.js:38]: I was considering this as well! Initially I wanted to put it in util, but then I realized that the behavior captured by this function isn't consistent across its usage. For example, consider [this line](https://github.com/MarkBind/markbind/pull/2239/files#diff-71dd0a800e64246506fda968bb42c6eccf3602da7963774c4bb42073cdd26509R19) in `build` that logs the message of the error and sets the exit code to 1, and also consider [this line](https://github.com/MarkBind/markbind/pull/2239/files#diff-ad9481704054365dffc7524f9956f377b4385a0968e3e037fa991bfcfcff8b37R45) in `serve` that directly logs a string and sets the exit code to 1. Previously, for the second case, it was calling the `handleError` function but creating a new `Error` with this message and then passing it into this function as can be seen [here](https://github.com/MarkBind/markbind/pull/2239/files#diff-e5ae170bb809d8234fc842e67b0309c8a179b5f9c4a30e0deefea928f42dd4fbL111), which I felt was an unnecessary step that can be removed.

In addition, this function just has two lines, so refactoring it out wouldn't change the readability too much (in fact, one could argue that refactoring out just two lines of code into another file could obfuscate readability). If anything, I think that refactoring out this function leads to unnecessarily rigid process for handling errors (`handleError` will always take in an `Error` object, display the message, and set exit code to 1), which shouldn't be the case.

Comment (yucheng11122017) [reply packages/cli/index.js:38]: Oh yes that true. Sounds good thanks for the response!

**Comment by raysonkoh** [comment]
Thanks for the refactoring work @lhw-1! It would be good if you could preserve the git history of those functions that you refactored into `cmd/`.

Maybe this would help: https://stackoverflow.com/questions/38729453/how-to-preserve-git-history-when-refactoring-into-multiple-files

**Comment by lhw-1** [comment]
Hi @raysonkoh, thanks for the guide! I've modified the files so that the git history is preserved, and as per the instructions on the link, running this command will show the git history:

```
git log --follow --find-copies packages/cli/src/cmd/build.js
```

However, when I view the individual file through GitHub (Files changed -> go to the individual file -> View file), I can't seem to view the full history. Is this fine, or should I revert the changes and re-start with the refactoring (with preserved git history)? I'm not familiar on whether this is a GitHub-side issue (and history will be preserved after the merge), or if the earlier commits are interfering with preserving the git history.

**Comment by raysonkoh** [comment]
I think it is a Github-side issue. Seems like they don't really have the feature where they are able to show that the file was copied from somewhere. Anyway I am able to see that the git history is preserved so it should be fine 👍 !

**Comment by jovyntls** [comment]
Hey @lhw-1 , for ease of the code review and sanity check, could I check if you've modified any of the functions (besides migrating from one file to the other)?

File: packages/cli/src/cmd/serve.js
Comment (jovyntls) [inline packages/cli/src/cmd/serve.js:62]: Since this method is only used once, can we revert it back to a constant config? I think it could be clearer as a constant variable that is defined where it is used.

**Review by jovyntls — CHANGES_REQUESTED** [review]
Just one comment! Could I check that these changes you've made, besides the migrating of code to different files?
* extract `addHandler` to a function
* extract `changeHandler` to a function
* extract `lazyReloadMiddleware` to a function
* extract `generateServerConfig` to a function
* extract `removeHandler` to a function

**Comment by lhw-1** [comment]
> Hey @lhw-1 , for ease of the code review and sanity check, could I check if you've modified any of the functions (besides migrating from one file to the other, and handleError as mentioned above)?

> Just one comment! Could I check that these changes you've made, besides the migrating of code to different files?
> 
> * extract `addHandler` to a function
> 
> * extract `changeHandler` to a function
> 
> * extract `lazyReloadMiddleware` to a function
> 
> * extract `generateServerConfig` to a function
> 
> * extract `removeHandler` to a function

Hi @jovyntls, thanks for reviewing the PR! To clarify, apart from `handleError` (and hence the error handling in various sections), I have not modified any of the functions. 

The five functions you mentioned above, as well as `syncOpenedPages`, were already themselves functions, just that they were declared within the `serve` command, as can be seen [here](https://github.com/MarkBind/markbind/blob/master/packages/cli/index.js#L221-L255) for the `lazyReloadMiddleware` function, and as for the rest, [here](https://github.com/MarkBind/markbind/blob/master/packages/cli/index.js#L137-L202). This is also primarily the reason why I put them inside the `serveUtil.js` file instead of the `cliUtil.js`.

File: packages/cli/src/cmd/serve.js
Comment (lhw-1) [reply packages/cli/src/cmd/serve.js]: Thanks for the good suggestion! Updated in [cb7dc59](https://github.com/MarkBind/markbind/pull/2239/commits/cb7dc59a94894f2c9ab995d975891fdda9d54f7c).

**Review by jovyntls — APPROVED** [review]
LGTM! 

> Additionally, I am curious on whether the naming of the new directory architecture, src/cmd/XXX.js - specifically, cmd - makes sense.

I haven't found any similar repos to use as reference, but `/cmd/` looks fine to me - can't think of any immediate issues with it :)

**Review by itsyme — APPROVED** [review]
Thanks for the work @lhw-1! LGTM! I have tested out the 4 commands and they seem to work well for me!

**Pull request merged**
